Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialCanvasDocument.cpp

```diff
@@ -352,4 +372,337 @@ namespace MaterialCanvas
+                        {
+                            auto sourceSlot = connection->GetSourceSlot();
+                            if (sourceSlot && sourceSlot->GetSlotDirection() == GraphModel::SlotDirection::Output)
+                            {
+                                replaceStringsInVec(
+                                    "SLOTVALUE",
+                                    AZStd::string::format("node%u_%s", sourceSlot->GetParentNode()->GetId(), sourceSlot->GetName().c_str()),
+                                    instructionsForSlot);
+                                break;
+                            }
+                        }
+                    }
+
+                    updateAndAddSlotInstructions(node, slot, instructionsForSlot);
+                }
+
+                for (const auto& slotConfig : nodeConfig.m_outputSlots)
+                {
+                    collectCommonSettings(slotConfig.m_settings);
+
+                    // If an output slot has any connections assume that we're going to use any configured instructions
+                    auto slot = dynamicNode->GetSlot(slotConfig.m_name);
+                    if (slot && !slot->GetConnections().empty())
+                    {
+                        AZStd::vector<AZStd::string> instructionsForSlot;
+                        collectSettingsAsVec(slotConfig.m_settings, "instructions", instructionsForSlot);
+                        updateAndAddSlotInstructions(node, slot, instructionsForSlot);
+                    }
+                }
+            }
+        }
+
+        AZ_TracePrintf("MaterialCanvasDocument", "Dumping data scraped from traversing material graph.\n");
+
+        for (const auto& node : nodes)
+        {
+            AZ_TracePrintf("MaterialCanvasDocument", "Sorted node: %s\n", node->GetTitle());
+        }
+
+        AZStd::string documentName;
+        AZ::StringFunc::Path::GetFullFileName(m_absolutePath.c_str(), documentName);
+        AZ::StringFunc::Replace(documentName, ".materialcanvas.azasset", "");
+
+        for (const auto& templatePath : templatePaths)
+        {
+            AZ_TracePrintf("MaterialCanvasDocument", "templatePath: %s\n", templatePath.c_str());
+            const AZStd::string resolvedPath = AtomToolsFramework::GetPathWithoutAlias(templatePath);
+            if(auto result = AZ::Utils::ReadFile(resolvedPath))
+            {
+                AZStd::string outputFullName;
+                AZ::StringFunc::Path::GetFullFileName(resolvedPath.c_str(), outputFullName);
+                AZStd::string outputPath = m_absolutePath;
+                AZ::StringFunc::Path::ReplaceFullName(outputPath, outputFullName.c_str());
+                AZ::StringFunc::Replace(outputPath, "MaterialGraphName", documentName.c_str());
+                AZ::StringFunc::Replace(outputPath, ".template", "");
+
+                AZStd::vector<AZStd::string> lines;
+                AZ::StringFunc::Tokenize(result.GetValue(), lines, '\n', true, true);
```

Comment:
May not be a bad idea to cache the line representation of the templates

Reply:
Do you mean loaded once and save it so that we can reuse it every time it recompiles? That would be a good optimization. I would also have to add monitoring to reload the templates in case people are planning on changing templates while the editor is open.